### ISSUE TITLE
False positive for no-unused-vars when first argument in an array of arguments is unused

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#7838]
**Tell us about your environment**

* **ESLint Version:** 3.12.2
* **Node Version:** 7.3.0
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
root: true
extends: airbnb-base
```

**What did you do? Please include the actual source code causing the issue.**

```js
module.exports = (baz, [foo, bar]) => {
  console.log(bar);
};
```

**What did you expect to happen?**

I expected `foo` not to be marked as an unused variable, because it pads `bar`. Removing `foo` would alter the behaviour of the code.

**What actually happened? Please include the actual, raw output from ESLint.**

`foo` got marked as an unused variable.

```
$ ./node_modules/.bin/eslint example.js 

/home/remco/d-centralize/appsemble-components/appsemble-build/example.js
  1:25  error    'foo' is defined but never used  no-unused-vars
  2:3   warning  Unexpected console statement     no-console

✖ 2 problems (1 error, 1 warning)
```

**Comment by vitorbal** [comment]
Thanks for the report. I was able to reproduce this in the demo page.
It looks like the `{ "args": "after-used" }` option doesn't play well with destructuring the function parameters.

**Comment by vitorbal** [comment]
Given you could do something like the following, I'm not entirely sure this should be considered a bug or not:
```javascript
module.exports = (baz, [, bar]) => {
  console.log(bar);
};
```

**Comment by remcohaszing** [comment]
@vitorbal I wasn't aware this syntax exists, but it makes sense.

Perhaps this could be added as an example to the documentation.

**Comment by platinumazure** [comment]
I agree we can/should improve the documentation here. @remcohaszing Would you be interested in submitting a pull request?

**Comment by vitorbal** [comment]
@remcohaszing sure, sounds like a good idea! Would you like to send a PR for that? I would imagine the new examples would go in the "after-used" section: http://eslint.org/docs/rules/no-unused-vars#args-after-used

**Comment by kaicataldo** [comment]
Agreed that it shouldn't be in the `after-used` section, since that specifically is for function arguments. The rule warns on destructured assignments too, so if we were to make this configurable it would have to be behind a different option.

That being said, given that we have `args: 'after-used'`, I do think it would make sense to match that with destructured array patterns as well.

**Comment by kaicataldo** [comment]
I think we should go ahead and update the docs with the open PR, though, and then continue the discussion over whether we want to add an option or not.

**Comment by not-an-aardvark** [comment]
I don't think it's necessary to add a new option for this case. `args: 'after-used'` is different because there is no way to have a "placeholder argument" without assigning it to an identifier, but with array destructuring, it is possible to have a placeholder, so we should encourage people to use that instead.

**Comment by kaicataldo** [comment]
I'm fine with not adding the option right now. We can evaluate again if it's requested.
